### PR TITLE
Support for custom configurations

### PR DESCRIPTION
**Pull request by johnnycheng0210**
Follow up more up to date rebase of this https://github.com/microsoft/typescript-go/pull/1636.

Address the feat request in this [ticket](https://github.com/microsoft/typescript-go/issues/945), by allowing custom configs to be passed to the tsgo LSP.

## Testing

- Load `.ts` file
- `findAllReferences` -> returns correct amount of references for this singular project
- update `"typescript.native-preview.customConfigFileName": "tsconfig.all.json"` to `solution` style project config -> returns all references in the entire repo (correct behavior)
- `tsconfig.all.json` is essentially `"include": ["./**/*"]` at root of our project.

### CODE REVIEW

File: _extension/src/client.ts [hunk 1]
```diff
@@ -90,10 +90,11 @@ export class Client {
         this.exe = exe;
         this.outputChannel.appendLine(`Resolved to ${this.exe.path}`);
 
-        // Get pprofDir
+		// Get pprofDir
```
Comment (DanielRosenwasser) [inline _extension/src/client.ts]: Switch this and every other TS line back to spaces

File: _extension/src/client.ts [hunk 2]
```diff
@@ -108,12 +109,17 @@ export class Client {
...
-            this.clientOptions,
-        );
+		this.client = new LanguageClient(
+			"typescript.native-preview",
+			"typescript.native-preview-lsp",
+			serverOptions,
+			{
+				...this.clientOptions,
+				initializationOptions: {
+					customConfigFileName,
```
Comment (DanielRosenwasser) [inline _extension/src/client.ts]: Rather than be part of initialization, we think this sort of thing would belong more on user preferences ([see here](https://github.com/microsoft/typescript-go/pull/1729).

Whether we just say "you need to restart the server" or we track this sort of thing on the server in some way is a bigger question.

**Review by DanielRosenwasser — DISMISSED** [review]
We haven't discussed this more broadly, but custom config specification has been a long-standing request with a lot of interest. The scope of implementing could grow a bit from what's here if we need something beyond a naive restart, but maybe this is good to start. I'll let others weigh in.

**Comment by jakebailey** [comment]
I'm not sure what happened with your PRs, but can you keep the one you intend us to look at open and keep pushing to that one?

**Comment by johnnycheng0210** [comment]
ah sorry - i was rebasing PR's - i will close the previous PR and use this one instead

File: _extension/src/client.ts (hunk 2 above)
Comment (johnnycheng0210) [reply _extension/src/client.ts]: hey @DanielRosenwasser i've just made the changes to have it be included in user preferences and have tested it works - could you take another look please

**Comment by jakebailey** [comment]
One thing this might be missing is some handling of on-change events; `(*Session).Configure` handles things; maybe `pendingConfigChanges` is sufficient to catch this?

File: _extension/package.json [hunk 3]
```diff
@@ -31,6 +31,12 @@
             {
                 "title": "TypeScript Native Preview",
                 "properties": {
+                    "typescript.native-preview.customConfigFileName": {
+                        "type": "string",
+                        "default": "",
+                        "description": "Custom config file name to use, before defaulting to tsconfig.json/jsconfig.json.",
```
Comment (jakebailey) [inline _extension/package.json]: An example value would be helpful here, like `tsconfig.editor.json`, just to show that we're looking for a _basename_, not a full path.

**Comment by jakebailey** [comment]
I tried this out and it seems to work. What I did to test it was to create `oops.ts` with:

```ts
declare const x: string | undefined;

const y: string = x;
```

Which fails normally as the default inferred project has strict=true.

Then I added `tsconfig.oops.json`:

```json
{
    "compilerOptions": {
        "strict": false
    },
    "files": [
        "oops.ts"
    ]
}
```

Then configured `tsconfig.oops.json` in the config. The diags refreshed and the error went away, so it did pick up the config.

Undoing the config didn't do anything, however, and it took a restart to get the diags to differ (even after modifying).

I'm not sure if that's good or bad, but it does seems suspicious.

**Comment by johnnycheng0210** [comment]
@jakebailey i did observe similar behavior with `findAllReferences` for example, when going from single project config -> singular solution project config -> single project config. The results after singular solution project -> single project config, didn't change. I thought it was to do with findAllReferences implementation but it appears its wider than that.

any clue why this might be happening? seems odd that one update works, but the subsequent one seems to remember the old ones config.

**Comment by johnnycheng0210** [comment]
I do wonder if a more naive approach, of requiring a window reload between custom config file name changes in user preferences is preferred. less likely to run into these edge cases, and still feels reasonably ergonomic?

**Comment by jakebailey** [comment]
I'm definitely confused, because when I write a test for this, the test seemingly does the right thing.

**Comment by johnnycheng0210** [comment]
hey @jakebailey any clue on this one? i've taken another look and am pretty confused too, it feels like its unrelated to the new config file name being updated, but moreso on the state of the language server when config files refresh? 🙏

**Comment by jakebailey** [comment]
I honestly have no idea ☹️

**Comment by johnnycheng0210** [comment]
> I do wonder if a more naive approach, of requiring a window reload between custom config file name changes in user preferences is preferred. less likely to run into these edge cases, and still feels reasonably ergonomic?

@jakebailey what do you think about this approach? when using this approach, we don't run into the above described edge case, albeit does feel less ergonomic.

**Comment by jakebailey** [comment]
I don't want to do that; we do not have anything that requires a client restart or a full window reload. I think we can figure out what's going on here. (I just have not personally had the chance to. Maybe @andrewbranch will know off the top of his head.)

**Comment by andrewbranch** [comment]
I'm not sure off the top of my head, I'd have to debug and step through.

**Comment by sspencer-canva** [comment]
This has been on hold for 3 weeks. It's a high-priority piece for us, and a number of engineers across multiple spaces are super keen to get it working here at Canva. Is there anything we can do to help? 🙏

**Comment by jakebailey** [comment]
This PR needs a merge from main at least (the snapshot code has changed a bit), but I still have not been able to explain the behavior I saw in testing. I'm trying to write some sort of test to show it.

**Comment by jakebailey** [comment]
I think I have a test and a fix. Will push it to the PR so we can get this moving.
